program/rust/src/accounts/price.rs

```diff
@@ -60,6 +62,95 @@ pub struct PriceAccount {
     pub comp_:              [PriceComponent; PC_COMP_SIZE as usize],
 }
 
+/// This is the new version of the PriceAccount.
+/// It will go live once we resize all the price accounts.
+/// It will include more publisher spots and will store cumulative sums to compute TWAPS.
+#[cfg(test)]
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct PriceAccountNew {
+    pub header:             AccountHeader,
+    /// Type of the price account
+    pub price_type:         u32,
+    /// Exponent for the published prices
+    pub exponent:           i32,
+    /// Current number of authorized publishers
+    pub num_:               u32,
+    /// Number of valid quotes for the last aggregation
+    pub num_qt_:            u32,
+    /// Last slot with a succesful aggregation (status : TRADING)
+    pub last_slot_:         u64,
+    /// Second to last slot where aggregation was attempted
+    pub valid_slot_:        u64,
+    /// Ema for price
+    pub twap_:              PriceEma,
+    /// Ema for confidence
+    pub twac_:              PriceEma,
+    /// Last time aggregation was attempted
+    pub timestamp_:         i64,
+    /// Minimum valid publisher quotes for a succesful aggregation
+    pub min_pub_:           u8,
+    pub unused_1_:          i8,
+    pub unused_2_:          i16,
+    pub unused_3_:          i32,
+    /// Corresponding product account
+    pub product_account:    Pubkey,
+    /// Next price account in the list
+    pub next_price_account: Pubkey,
+    /// Second to last slot where aggregation was succesful (i.e. status : TRADING)
+    pub prev_slot_:         u64,
+    /// Aggregate price at prev_slot_
+    pub prev_price_:        i64,
+    /// Confidence interval at prev_slot_
+    pub prev_conf_:         u64,
+    /// Timestamp of prev_slot_
+    pub prev_timestamp_:    i64,
+    /// Last attempted aggregate results
+    pub agg_:               PriceInfo,
+    /// Publishers' price components
+    pub comp_:              [PriceComponent; PC_COMP_SIZE as usize],
+    /// Cumulative sums of aggregative price and confidence used to compute arithmetic moving averages
+    pub price_cumulative:   PriceCumulative,
+}
+
+#[cfg(test)]
+impl PriceAccountNew {
+    /// This function gets triggered when there's a succesful aggregation and updates the cumulative sums
+    pub fn update_price_cumulative(&mut self) {
```

Comment:
aren't there some assumptions about trading and stuff that need to be satisfied in order to call this function?

IMO this function does so little it would be better to inline the call to update in the calling code (which also guarantees the assumptions are satisfied)

Reply:
The reason it's not inlined is some rust borrower annoyance that makes it not very clean.
I added a check that status == TRADING whenever this function gets called.